keps/sig-scheduling/20180409-scheduling-framework.md

```diff
@@ -198,49 +197,55 @@ post-filter extension point.
 
 ### Scoring
 
-These plugins are used to rank nodes that have passed the filtering phase. The
-scheduler will call each scoring plugin for each node. There will be a well
-defined range of integers representing the minimum and maximum scores. After the
-[normalize scoring](#normalize-scoring) phase, the scheduler will combine node
-scores from all plugins according to the configured plugin weights.
+These plugins have two phases:
 
-### Normalize scoring
-
-These plugins are used to modify scores before the scheduler computes a final
-ranking of Nodes. A plugin that registers for this extension point will be
-called with the [scoring](#scoring) results from the same plugin. This is called
-once per plugin per scheduling cycle.
+1. The first phase is called "score" which is used to rank nodes that have passed 
+the filtering phase. The scheduler will call `Score` of each scoring plugin for
+each node. There will be a well defined range of integers representing the minimum
+and maximum scores. 
+2. The second phase is "normalize scoring" which is used to modify scores before
+the scheduler computes a final ranking of Nodes, and each score plugin receives
+scores given by the same plugin to all nodes in "normalize scoring" phase. 
+`NormalizeScore` is called once per plugin per scheduling cycle right after 
+"score" phase. After the "normalize scoring" phase, the scheduler will combine 
+node scores from all plugins according to the configured plugin weights.
 
 For example, suppose a plugin `BlinkingLightScorer` ranks Nodes based on how
 many blinking lights they have.
 
 ```go
-func ScoreNode(_ *v1.Pod, n *v1.Node) (int, error) {
-   return getBlinkingLightCount(n)
+func (*BlinkingLightScorer) Score(pc *PluginContext, _ *v1.Pod, nodeName string) (int, *Status) {
+   return getBlinkingLightCount(nodeName)
 }
 ```
 
 However, the maximum count of blinking lights may be small compared to
-`NodeScoreMax`. To fix this, `BlinkingLightScorer` should also register for this
-extension point.
+`NodeScoreMax`. To fix this, `BlinkingLightScorer` should also implement `NormalizeScore`.
 
 ```go
-func NormalizeScores(scores map[string]int) {
+func (*BlinkingLightScorer) NormalizeScore(pc *PluginContext, _ *v1.Pod, scores NodeScoreList) *Status {
    highest := 0
    for _, score := range scores {
       highest = max(highest, score)
    }
    for node, score := range scores {
       scores[node] = score*NodeScoreMax/highest
    }
+   return nil
 }
 ```
 
-If any normalize-scoring plugin returns an error, the scheduling cycle is
-aborted.
+If `NormalizeScore` returns an error, the scheduling cycle is aborted. We could also
+implement `NormalizeScore` as an noop when it is not necessary to modify scores.
```

Comment:
nit: "modify scores" -> "normalize scores"

Currently there are two interfaces: `ScorePlugin` and `ScoreWithNormalizePlguin` which embeds `ScorePlugin`. User can choose to only implement `ScorePlugin` if no normalize is required. This is equivalent to implementing `NormalizeScore` as a noop.
But I think your phrasing is also fine. The doc doesn't need to be that detailed.

Reply:
> nit: "modify scores" -> "normalize scores"

"implement `NormalizeScore` as an noop when it is not necessary to normalize scores." seems odd, I prefer to go with "modify scores". :)

> the doc doesn't need to be that detailed.

+1

Reply:
hmm, I think our recommendation should be "NormalizeScore does not need to be implemented if not needed", not  "implement Normalize as a noop when it is not necessary".

Reply:
> hmm, I think our recommendation should be "NormalizeScore does not need to be implemented if not needed", not "implement Normalize as a noop when it is not necessary".

Does it depend on the final API of the scheduling framework?

```go
type ScorePlugin interface {
  Score()..
  NormalizeScore()...
}

or 

type ScorePlugin interface {
  Score()
  NormalizeScore()
}

type ScoreWithNormalizePlugin interface {
  Score()
}
```

Reply:
> > nit: "modify scores" -> "normalize scores"
> 
> "implement `NormalizeScore` as an noop when it is not necessary to normalize scores." seems odd, I prefer to go with "modify scores". :)

To me "modify scores" is too broad: "normalize" or "apply weight" both modifies scores. I prefer to go with more specifics.

Reply:
> > hmm, I think our recommendation should be "NormalizeScore does not need to be implemented if not needed", not "implement Normalize as a noop when it is not necessary".
> 
> Does it depend on the final API of the scheduling framework?
> 
> ```go
> type ScorePlugin interface {
>   Score()..
>   NormalizeScore()...
> }
> 
> or 
> 
> type ScorePlugin interface {
>   Score()
>   NormalizeScore()
> }
> 
> type ScoreWithNormalizePlugin interface {
>   Score()
> }
> ```

The interface we agreed on is:
```go
type ScorePlugin interface {
  Score()
}

type ScoreWithNormalizePlugin interface {
  ScorePlugin
  NormalizeScore()
}
```